File: 004/Problem004.py
''' 
  * Problem004.py
  *
  * Project Euler: Problem 4
  * 
  * Problem: Find the largest palindrome made from the product of two 3-digit numbers.
  *          
  *
'''
import time
start_time = time.time()

def is_palindrome(n):
    num = str(n)
    
    if len(num) % 2 == 0:
        max = len(num) / 2
    else:
        max = (len(num) - 1) / 2

    for i in range(max):
        if  num[i] == num[len(num) - 1 - i]:
            pali = True
        else:
            pali = False
            break
    
    return pali

for i in xrange(999,899,-1):
    for j in xrange(999,899,-1):
        if is_palindrome(i*j):
            print i*j
            break
    if is_palindrome(i*j):
        break

print("--- %s seconds ---" % (time.time() - start_time))

File: 004/Problem004.java
public class Problem004 {

    public static void main( String[] args ) {
        int result = largestPalindromeProduct(999,999);
        System.out.println(result);
    }

    private static int largestPalindromeProduct(int a, int b) {
        int largest = 0;
        String current;
        for (int i = 0; i < a; i++) {
            for (int j = i; j < b; j++) {
                current = String.valueOf(i*j);
                if (isPalindrome(current) && i*j > largest) {
                    largest = i*j;
                }
            }
        }
        return largest;
    }

    private static boolean isPalindrome(String s) {
        for (int i = 0; i < s.length()/2; i++) {
            if (s.charAt(i) != s.charAt(s.length() - i - 1)) {
                return false;
            }
        }
        return true;
    }
}
